ccmbioinfo/lfs
# Reduce webpack file sizes

To test:
- does the build feel a bit faster?
- does the app feel a bit faster?
- check that everything still works

Some observations:
- When A.js imports a function from B.js, webpack will consider that ALL of B.js and its recursive dependencies is needed by A.js, so when building A.js as an entrypoint, it will be really fat. It is best to keep small utility functions out of big components. For example, in this PR I moved the subject identifier code out of Subject.js because small entrypoints that just needed to display a subject identifier ended up bundling Subjects+Forms+Pedigree+Dicom, so most entrypoints ballooned to many megabytes in size. More work will be needed to reduce dependencies between entry points.
- Webpack5 introduced the dependOn feature, allowing entrypoints to not duplicate other entry points. However, in the version we were using before this was barely functional: it could work with a simple `A dependsOn B` declaration, but didn't support `A dependOn B + B dependOn C` or `A dependOn B and C`. In the latest webpack version this works quite well, supporting a complex graph of dependencies.
- In the old version of webpack , `eval(js file)` returned the module, but version 5.25.0 broke this behavior (intentionally). They really force you to use `import` (which gets translated to `__webpack__require__`), and there is no nice way to get something out of `eval`. It is possible using some of the `libraryTarget` options, if we pretend that the whole app is a library, but that's not semantically correct and is also prone to errors, since each entrypoint would be considered the same library, leading to overrides or race conditions. In the end I had to dig deep into webpack code to figure out a [nice and simple way to add back the `return` at the end of the generated js](https://github.com/data-team-uhn/cards/pull/1447/commits/c783a95671fe399b0552f15f5e903cbc48180a11#diff-cd9057fcd7bc38b9812edb8519872d7c022b412555d5f78e4e8a301d3291d99bR30-R41).
- Related to the point above, webpack is also really hard to be used for building a dynamic application that doesn't have a fixed structure. In theory, it is possible to do something like `import(someDynamicPathComputedAtRuntime)`, which is called lazy import, but in practice it still requires listing all the possible paths at compile time, the implementation doesn't really look up dynamically that module, but just checks which of the possible files found at compile time was requested. As such, it is only useful for something like "we'll need one of these 3 files, but we'll decide at runtime which of those 3 it is", and not something like "we'll allow any file to be plugged in and loaded". This means that we still need to use `eval(soure)` to dynamically load components from extension points, even though it is ugly.
- However, one webpack feature that may be worth investigating is [module federation](https://webpack.js.org/concepts/module-federation/), which does allow importing a module that was not available at compile time, so it's a promising feature for splitting the frontend back into individually compiled modules, and the codebase into a core repo and seprate smaller repos for each client project. 
- Recently, webpack has been working on adding support for compiling to native modules, which would not translate `import/export` code into its own `__webpack__require__` workaround, but there are still bugs in the implementation: sometimes it puts `"use strict"` at the beginning of the files, which is not allowed for modules, and in dev mode it adds a comment about using an eval devtool, which also isn't allowed. But perhaps in the future we'll be able to switch to native modules.
- Webpack optimizes how modules are loaded to only load each module once. That means that the first `eval(A.js)` returns the module, but the second `eval(A.js)` will not run the code and returns `undefined`.
- In development mode, the generated code ends up something like `this is module [A.js]`, but in production mode it is `this is module [A.js, and B.js, and C.js]` if all 3 are declared as entrypoints, and A.js happens to include B and C because it uses them and doesn't declare them as dependencies. Combined with the point above, this means that if `A.js` is loaded before `B.js`, it is possible to not be able to load `B.js` and get the components out of it, since `B.js` was already loaded from the `A.js` module. This means that dependencies **MUST** be declared, or else everything will work fine in development, but releases will crash.
- Tree shaking (removing unused code from compiled entrypoints) doesn't work because some of our code has side effects, e.g. all the registered components (answer types in forms, item types in the questionnaire editor). But we can manually annotate files that are free of side effects to let webpack prune away unused code.
- I've been using [Webpack Visualizer](https://chrisbateman.github.io/webpack-visualizer/) to analyze what ends up in each entrypoint.

Things to do:
- Reduce inter-module dependencies using the webpack analyzer, figure out why big things end up in little things, move small utility functions out of big components into separate files, add `dependsOn` where otherwise impossible to remove a dependency; I've added a lot of `dependsOn Forms` and `dependsOn Questionnaires` on entrypoints that don't really requires all of Forms just to quickly reduce the size, but these dependencies should be optimized
- `assetManager.js` needs to fetch the source for a module while its dependencies are loaded, not after; I had this code working, but it got lost during the last bugfix/refactoring.
- Annotate files without side effects with `/*#__PURE__*/` to [allow tree shaking](https://webpack.js.org/guides/tree-shaking/) in them
- Make Dicom an entry point to not have it bundled in `Forms.js`
- Really make answer types modular, to allow Dicom/Pedigree code to be skipped in projects that don't need those
- Since we have real data about which browsers are used, we should add the list of browsers we want to support to the build configuration, to use fewer shims and workarounds for the generated code.
